cbmc/proofs/mqtt_state.c

```diff
@@ -0,0 +1,556 @@
+#endif
+    IFF( pInfo->pPreviousSubscriptions == NULL,
+	 pInfo->previousSubscriptionCount == 0 ) &&
+    valid_IotMqttSubscriptionArray(pInfo->pPreviousSubscriptions,
+				   pInfo->previousSubscriptionCount);
+}
+
+/****************************************************************
+ * IotMqttSubscription
+ ****************************************************************/
+
+IotMqttSubscription_t *allocate_IotMqttSubscription(IotMqttSubscription_t *pSub)
+{
+  if ( pSub == NULL ) pSub = malloc_can_fail( sizeof( *pSub ) );
+  if ( pSub == NULL ) return NULL;
+
+  pSub->pTopicFilter = malloc_can_fail( pSub->topicFilterLength );
+  return pSub;
+}
+
+bool valid_IotMqttSubscription(IotMqttSubscription_t *pSub)
+{
+  return
+    pSub &&
+
+    VALID_STRING( pSub->pTopicFilter,  pSub->topicFilterLength ) &&
+    VALID_CBMC_SIZE( pSub->topicFilterLength )
+
+#ifdef TOPIC_LENGTH_MAX
+    && pSub->topicFilterLength < TOPIC_LENGTH_MAX
+#endif
+    ;
+}
+
+/****************************************************************
+ * IotMqttSubscription array list
+ ****************************************************************/
+
+IotMqttSubscription_t *allocate_IotMqttSubscriptionArray(IotMqttSubscription_t *pSub, size_t length)
+{
+  if (pSub == NULL ) pSub = malloc_can_fail( length * sizeof( *pSub ) );
+  if (pSub == NULL ) return NULL;
+
+  for (int i = 0; i < length; i++)
+    allocate_IotMqttSubscription(pSub + i);
+
+  return pSub;
+}
+
+bool valid_IotMqttSubscriptionArray(IotMqttSubscription_t *pSub, size_t length)
+{
+  if ( !IFF(pSub == NULL, length == 0 ) ) return 0;
+  if ( pSub == NULL ) return 0;
+
+  bool result = 1;
+  for (int i = 0; i < length; i++)
+    result = result && valid_IotMqttSubscription(pSub + i);
+
+  return
+#ifdef SUBSCRIPTION_COUNT_MAX
+    length < SUBSCRIPTION_COUNT_MAX &&
+#endif
+    result;
+}
+
+/****************************************************************
+ * IotMqttSubscription array list
+ ****************************************************************/
+
+// Subscription linked list elements
+
+_mqttSubscription_t *allocate_IotMqttSubscriptionListElt(_mqttSubscription_t *pElt)
+{
+  size_t length;
+  // Assumption avoids arithmetic overflow in malloc, rechecked in valid_*
+  __CPROVER_assume( length <
+		    CBMC_MAX_OBJECT_SIZE - sizeof( _mqttSubscription_t ) );
+
+  if ( pElt == NULL ) pElt = malloc_can_fail( sizeof( *pElt ) + length);
+  if ( pElt == NULL ) return NULL;
+
+  pElt->link.pPrevious = NULL;
+  pElt->link.pNext = NULL;
+  pElt->topicFilterLength = length;
+  return pElt;
+}
+
+bool valid_IotMqttSubscriptionListElt(_mqttSubscription_t *pElt)
+{
+  if ( pElt == NULL ) return 0;
+
+  return
+#ifdef TOPIC_LENGTH_MAX
+    pElt->topicFilterLength < TOPIC_LENGTH_MAX &&
+#endif
+    pElt->topicFilterLength < CBMC_MAX_OBJECT_SIZE - sizeof( _mqttSubscription_t ) &&
+    __CPROVER_r_ok(pElt->pTopicFilter, pElt->topicFilterLength) &&
+    __CPROVER_w_ok(pElt->pTopicFilter, pElt->topicFilterLength);
```

Comment:
`w_ok` and `r_ok` are currently synonyms in CBMC.  Tho they might not be forever.